web/blog/tenzir-node-v4.24/index.md

```diff
@@ -0,0 +1,221 @@
+```
+
+```tql
+{s: "1,2;3", xs: ["1", "2", "3"]}
+```
+
+```tql title="Join a list"
+from {xs: ["1", "2", "3"]}
+s = xs.join(",")
+```
+
+```tql
+{xs: ["1", "2", "3"], s: "1,2,3"}
+```
+
+### Modify and Remove List Elements
+
+Two new functions `map` and `where` allow you to modify and filter lists.
+
+```tql title="Modify list elements"
+from {xs: [1, 2, 3]}
+doubled = xs.map(x, x * 2)
+```
+
+```tql
+{xs: [1, 2, 3], doubled: [2, 4, 6]}
+```
+
+```tql title="Remove list elements"
+from {xs: [1, 2, 3, 4, 5]}
+large = xs.where(x, x > 3)
+```
+
+```tql
+{xs: [1, 2, 3, 4, 5], large: [4, 5]}
+```
+
+### Aggregation Functions
+
+With the `summarize` operator, Tenzir already supports aggregating values across
+multiple events. Now, all aggregation functions work on lists as well as regular
+functions:
+
+```tql title="Sum of a list"
+from {xs: [1, 2, 3]}
+total = xs.sum()
+```
+
+```tql
+{xs: [1, 2, 3], total: 6}
+```
+
+```tql title="Minimum of a list"
+from {xs: [1, 2, 3]}
+lowest = xs.min()
+```
+
+```tql
+{xs: [1, 2, 3], lowest: 1}
+```
+
+```tql title="Distinct values in a list"
+from {xs: [1, 2, 2, 3]}
+unique = xs.distinct()
+```
+
+```tql
+{xs: [1, 2, 2, 3], unique: [1, 2, 3]}
+```
+
+:::info Aggregation Functions
+With this change, all aggregation functions work with lists. Check the
+[functions reference](/tql2/functions#aggregation) for a full list of available
+functions.
+:::
+
+## Contexts in TQL2
+
+Contexts have arrived in TQL2, bringing powerful enrichment capabilities to your
+pipelines. Our new [Enrichments](/next/enrichments) documentation explains how
+they work in detail, but here's a quick overview of what you can do with
+contexts.
+
+Contexts are stateful objects that allow you to add contextual data to your
+events. You can use them to:
+
+- Build lookup tables for fast IP-to-asset mapping
+- Create Bloom filters for efficient membership testing of large sets
+- Leverage GeoIP databases for geographic IP enrichment
+
+Here's a simple example of using a lookup table context:
+
+```tql title="Create a lookup table context"
+context::create_lookup_table "ip-to-hostname"
+```
+
+```tql title="Populate lookup table with data"
+subscribe "suricata"
+where event_type == "dns"
+unroll dns.answers
+where dns.answers.rrtype in ["A", "AAAA"]
+context::update "my_threats", key: dest_ip, dns.answers.rdata
+```
+
+```tql title="Enrich other data with context"
+subscribe "zeek"
+context::enrich "ip-to-hostname", key=id.resp_h, into=hostname
+…
+```
+
+Each context type has its own strengths:
+
+- **Lookup Tables** excel at key-value mappings with features like subnet
+  matching and value aggregation
+- **Bloom Filters** provide space-efficient set membership testing
+- **GeoIP Databases** offer specialized geographic information lookup for IP
+  addresses
+
+We've made contexts first-class citizens in TQL2 with robust APIs for creation,
+updates, and enrichment operations. At this point, only the `lookup` operator is
+missing in TQL2.
```

Comment:
I would remove the `lookup` part, as you're also describing below what's missing. Once is enough.

Reply:
Good call. I removed the entire paragraph, as the rest of it didn't make much sense without this.